samples/secure_tunneling/secure_tunnel/main.cpp

```diff
@@ -0,0 +1,278 @@
+                    fprintf(stdout, "Send Data Complete in Destination Mode\n");
+                }
+                else
+                {
+                    fprintf(stderr, "Send Data Failed with error code %d\n", error_code);
+                }
+
+                break;
+            case AWS_SECURE_TUNNELING_SOURCE_MODE:
+                if (!error_code)
+                {
+                    fprintf(stdout, "Send Data Complete in Source Mode\n");
+                }
+                else
+                {
+                    fprintf(stderr, "Send Data Failed with error code %d\n", error_code);
+                }
+                break;
+        }
+    };
+
+    auto OnDataReceive = [&](const struct aws_byte_buf &data) {
+        receivedData = std::string((char *)data.buffer, data.len);
+        string returnMessage = "Echo:" + receivedData;
+
+        fprintf(stdout, "Received: \"%s\"\n", receivedData.c_str());
+
+        switch (localProxyMode)
+        {
+            case AWS_SECURE_TUNNELING_DESTINATION_MODE:
+                fprintf(stdout, "Data Receive Complete in Destination\n");
+                fprintf(stdout, "Sending response message:\"%s\"\n", returnMessage.c_str());
+                mSecureTunnel->SendData(ByteCursorFromCString(returnMessage.c_str()));
+                break;
+            case AWS_SECURE_TUNNELING_SOURCE_MODE:
+                fprintf(stdout, "Data Receive Complete in Source\n");
+                break;
+        }
+    };
+
+    /*
+     * This only fires in Destination Mode
+     */
+    auto OnStreamStart = [&]() { fprintf(stdout, "Stream Started in Destination Mode\n"); };
+
+    auto OnStreamReset = [&]() { fprintf(stdout, "Stream Reset\n"); };
+
+    auto OnSessionReset = [&]() { fprintf(stdout, "Session Reset\n"); };
+
+    /*********************** Secure Tunnel Setup ***************************/
+
+    /*
+     * Create a new SecureTunnel using the SecureTunnelBuilder
+     */
+    mSecureTunnel =
+        SecureTunnelBuilder(
+            Aws::Crt::g_allocator, bootstrap, SocketOptions(), accessToken.c_str(), localProxyMode, endpoint.c_str())
+            .WithRootCa(caFile.c_str())
+            .WithOnConnectionComplete(OnConnectionComplete)
+            .WithOnConnectionShutdown(OnConnectionShutdown)
+            .WithOnSendDataComplete(OnSendDataComplete)
+            .WithOnDataReceive(OnDataReceive)
+            .WithOnStreamStart(OnStreamStart)
+            .WithOnStreamReset(OnStreamReset)
+            .WithOnSessionReset(OnSessionReset)
+            .Build();
+
+    if (!mSecureTunnel)
+    {
+        fprintf(stderr, "Secure Tunnel Creation failed");
+        exit(-1);
+    }
+
+    mSecureTunnel->Connect();
+    int messageCount = 0;
+
+    while (1)
+    {
+        std::this_thread::sleep_for(3000ms);
+        if (localProxyMode == AWS_SECURE_TUNNELING_SOURCE_MODE)
+        {
+            messageCount++;
+            string toSend = to_string(messageCount) + ": " + message.c_str();
```

Comment:
So ... `mSecureTunnel->Connect()` that gets called above is async. Even if it returns AWS_OP_SUCCESS, the connection still hasn't succeeded, it just means the connection attempt successfully *started*.

You need to wait until the connection is successful before trying to send data. This is succeeding right now because it waits 3 seconds before sending data, and you happen to connect in those 3 seconds, but that won't always be the case..

You probably want to do something like the MQTT samples, that use synchronization primitives to halt the main thread until the connection callback fires. Look at how [connectionCompletedPromise](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/2d8447ccb2c4a4a89b504a382c5329c8181140a1/samples/mqtt/basic_pub_sub/main.cpp#L434) is used, the main thread waits on `connectionCompletedPromise` to know when the connection attempt completes, and the callbacks set `connectionCompletedPromise` to indicate whether it succeeded or failed